Flush messages before DeathWatchNotification, #28695

patriknw opened the pull request:
* Since DeathWatchNotification is sent over the control channel it may overtake
  other messages that have been sent from the same actor before it stopped.
* It can be confusing that Terminated can't be used as an end-of-conversation marker.
* In classic Remoting we didn't have this problem because all messages were sent over
  the same connection.

Drawback is longer latency for the Terminated notification under normal circumstances.

A few remaining things but ready for first review and discussion.

I would also like to investigate how we cope with many actors terminating at the same time (for example when ActorSystem is shutdown). This may cause additional load for that scenario.

References #28695

patriknw committed:
Flush messages before DeathWatchNotification, #28695

* Since DeathWatchNotification is sent over the control channel it may overtake
  other messages that have been sent from the same actor before it stopped.
* It can be confusing that Terminated can't be used as an end-of-conversation marker.
* In classic Remoting we didn't have this problem because all messages were sent over
  the same connection.

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -248,79 +244,6 @@ private[remote] trait OutboundContext {
 
 }
 
-/**
```
Moved to separate file.

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/FlushOnShutdown.scala:
```diff
@@ -0,0 +1,83 @@
+ * INTERNAL API
+ */
+@InternalApi
+private[remote] object FlushOnShutdown {
```
This was moved from other place to separate file. No changes.

patriknw commented on akka-remote/src/test/scala/akka/remote/artery/DeathWatchNotificationSpec.scala:
```diff
@@ -0,0 +1,111 @@
+    receiverProbe.expectMsg("echo")
+
+    sender ! PoisonPill
+    receiverProbe.receiveN(messages.size).toVector shouldBe messages
```
This test fails every time without the flushing. Often the Terminated arrives as the second message.

hepin1989 commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -924,10 +847,30 @@ private[remote] abstract class ArteryTransport(_system: ExtendedActorSystem, _pr
+  // Purpose of this stage is flushing, the sender can wait for the ACKs up to try flushing
+  // pending messages.
+  val flushReplier: Flow[InboundEnvelope, InboundEnvelope, NotUsed] = {
+    Flow[InboundEnvelope].filter { envelope =>
```
It's not a pure predicate function now

johanandren commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -924,10 +847,30 @@ private[remote] abstract class ArteryTransport(_system: ExtendedActorSystem, _pr
+        case _ => true
+      }
+    }
+  }
```
So this runs inside the regular and control inbound, acking and going through the outgoing regular with the ack, that still doesn't guarantee the Terminated will come after all messages though, does it, just makes it much much less likely?

patriknw replied:
I don't think we have any async boundaries after this so when reply is sent from one lane the preceding message has already been delivered to the mailbox of the destination actor. Is that thinking correct?

johanandren replied:
Not sure that is a guarantee, but extremely likely, then adding the network hop of the ack on top the message should have ended up in the inbox...

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/FlushBeforeDeathWatchNotification.scala:
```diff
@@ -0,0 +1,79 @@
+        context.stop(self)
+    case Timeout =>
+      log.debug("Flush timeout, [{}] remaining", remaining)
+      context.stop(self)
```
My thinking is that then it's pretty much as if the other ordinary messages were lost. It's not a guarantee, but a best effort to deliver other messages first.

johanandren reviewed (commented):
I think this is reasonable, but also a bit worried about thundering herd like problem on shutdown (or perhaps shard rebalance) like you mention.

patriknw commented:
I did some experiments in a test to understand (remind myself) how the RemoteWatcher/ClusterWatcher works.

There is one optimization in place to support many watchers. It doesn't matter how many watchers that are watching the same watchee. The RemoteWatcher takes over the responsibility from the watcher. So there will be a watch link between the RemoteWatcher and each watchee.

When the watchees are terminated they will all send DeathWatchNotification to the RemoteWatcher on the watcher side. This is where we can have a problem that is increased by the flush. Especially when the ActorSystem is shutdown that will trigger many DeathWatchNotification messages.

When there is termination in progress it would be enough with one single DeathWatchNotification telling that the entire system has been terminated. (Maybe implemented by that the RemoteWatchers watch each other).

We could have a similar thing for when the watcher side is terminating. Now the RemoteWatcher will send Unwatch to all watchees. That can be replaced by a single message.

patriknw committed:
don't send DeathWatchNotification when system is terminating

* when using Cluster we can rely on that the other side will publish AddressTerminated
  when the member has been removed
* it's actually already a race condition that often will result in that the DeathWatchNotification
  from the terminating side
  * in DeathWatch.scala it will remove the watchedBy when receiving AddressTerminated, and that
    may (sometimes) happen before tellWatchersWeDied

patriknw committed:
name counter

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/Association.scala:
```diff
@@ -367,11 +378,20 @@ private[remote] class Association(
+              s"flush-${UUID.randomUUID()}")
+            implicit val ec = materializer.executionContext
+            flushingPromise.future.onComplete { _ =>
+              sendSystemMessage(outboundEnvelope)
```
yes, better to send it than drop it

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/Association.scala:
```diff
@@ -367,11 +378,20 @@ private[remote] class Association(
+              FlushBeforeDeathWatchNotification
+                .props(flushingPromise, settings.Advanced.ShutdownFlushTimeout, this)
+                .withDispatcher(Dispatchers.InternalDispatcherId),
+              s"flush-${UUID.randomUUID()}")
```
fixed

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/Association.scala:
```diff
@@ -352,6 +356,26 @@ private[remote] class Association(
+
+    def sendSystemMessage(outboundEnvelope: OutboundEnvelope): Unit = {
+      outboundEnvelope.message match {
+        case d: DeathWatchNotification if !shouldSendDeathWatchNotification(d) =>
```
Added this, so it's not sending the DeathWatchNotification if using Cluster and the actor system is terminating. The other side will be notified by AddressTerminated when member has been removed.

I actually found that sometimes it's not sending the DeathWatchNotification, also before this change. We were probably not fully aware of that but it makes me more comfortable with the change. When ClusterRemoteWatcher sees MemberRemoved it will publish AddressTerminated. So that will happen for all other members in the cluster when the ActorSystem is terminating. In DeathWatch.scala that will be received and the `watchedBy` will be removed (sometimes) before `tellWatchersWeDied`. In that case DeathWatchNotification message will not be sent to the watchers.

https://github.com/akka/akka/blob/9641780054bc587139fd348739aee750a7016941/akka-actor/src/main/scala/akka/actor/dungeon/DeathWatch.scala#L199-L203

patriknw committed:
fixme

patriknw commented:
Still todo: We could have a similar thing for when the watcher side is terminating. Now the RemoteWatcher will send Unwatch to all watchees (I think).

patriknw committed:
same for Unwatch

* to avoid sending many Unwatch messages when watcher's ActorSystem is terminated
* same race exists for Unwatch as for DeathWatchNotification, if RemoteWatcher publishAddressTerminated
  before the watcher is terminated

patriknw committed:
unused import

patriknw commented:
added same for Unwatch
* to avoid sending many Unwatch messages when watcher's ActorSystem is terminated
* same race exists for Unwatch as for DeathWatchNotification, if RemoteWatcher publishAddressTerminated
  before the watcher is terminated

patriknw commented:
This is ready for final review.

patriknw committed:
config for the flush timeout, and possibility to disable

johanandren commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -980,6 +980,12 @@ akka {
 
+        # Before sending notificaiton of terminated actor (DeathWatchNotification) other messages
+        # will be flushed to make sure that the Terminated message arrives after other messages.
+        # It will wait this long for the flush acknoledgement before continuing.
```
acknowledgement

johanandren commented on akka-remote/src/main/scala/akka/remote/artery/ArterySettings.scala:
```diff
@@ -152,7 +152,18 @@ private[akka] final class ArterySettings private (config: Config) {
+            .getMillisDuration("death-watch-notification-flush-timeout")
+            .requiring(
+              interval => interval > Duration.Zero,
+              "death-watch-notification-flush-timeout must be more than zero")
```
"must be either more than zero or 'off'"

johanandren commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -925,10 +848,30 @@ private[remote] abstract class ArteryTransport(_system: ExtendedActorSystem, _pr
+            case OptionVal.Some(snd) =>
+              snd.tell(FlushAck, ActorRef.noSender)
+            case OptionVal.None =>
+              log.error("Expected sender for Flush message")
```
Some more details from `envelope` in this error log to help narrow it down in case it happens?

johanandren commented on akka-remote/src/main/scala/akka/remote/artery/Control.scala:
```diff
@@ -43,12 +43,22 @@ private[remote] final case class Quarantined(from: UniqueAddress, to: UniqueAddr
+/**
+ * INTERNAL API
+ */
+private[remote] case object FlushAck extends ArteryMessage
```
Missing `@InternalApi`

johanandren reviewed (approved):
LGTM with a few nitpicks

patriknw committed:
fix test

patriknw committed:
feedback

chbatey commented on akka-remote/src/main/scala/akka/remote/artery/Association.scala:
```diff
@@ -368,11 +401,18 @@ private[remote] class Association(
       try {
         val outboundEnvelope = createOutboundEnvelope()
         message match {
+          case d: DeathWatchNotification if deathWatchNotificationFlushEnabled && shouldSendDeathWatchNotification(d) =>
```
i guess this means system. messages will be re-ordered but i don't think that matters

patriknw replied:
For different sender-receiver pairs it never matters. DeathWatchNotification is always the last message anyway, as far as I can see.

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/FlushBeforeDeathWatchNotification.scala:
```diff
@@ -0,0 +1,83 @@
+
+  override def preStart(): Unit = {
+    try {
+      remaining = association.sendFlush(self)
```
that should not be needed, I can look into filtering out the control queue

patriknw replied:
excluded the control queue from in the flushing: 496bca01e8

patriknw committed:
exclude control queue from flush

chbatey reviewed (approved):
LGTM

raboof commented:
Suspicious that the `DeathWatchNotificationSpec` failed just after merging this, see #29349
